File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
const findTarget = (root, k) => {
    const set = new Set();
    const dfs = (root, k) => {
        if (!root) return false;
        const complement = k - root.val;
        if (set.has(complement)) return true;
        set.add(root.val);
        return dfs(root.left, k) || dfs(root.right, k);
    }
    return dfs(root, k);
};

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        vals = set()
        def dfs(root):
            if root == None:
                return False
            comp = k - root.val
            if comp in vals:
                return True
            else:
                vals.add(root.val)
                return dfs(root.left) or dfs(root.right)
        return dfs(root)